From 8c58a99221415ca7c3d5ce50dcffefa14e421928 Mon Sep 17 00:00:00 2001
From: Tomas Orsava <torsava@redhat.com>
Date: Tue, 12 Nov 2019 17:24:20 +0100
Subject: [PATCH] Subject: Prevent removing of the system packages installed
 under /usr/lib

when pip install -U is executed.

Resolves: rhbz#1550368

Co-Authored-By: Michal Cyprian <m.cyprian@gmail.com>
Co-Authored-By: Victor Stinner <vstinner@redhat.com>
---
 src/pip/_internal/legacy_resolve.py  |  5 ++++-
 src/pip/_internal/req/req_install.py |  3 ++-
 src/pip/_internal/utils/misc.py      | 11 +++++++++++
 3 files changed, 17 insertions(+), 2 deletions(-)

diff --git a/src/pip/_internal/legacy_resolve.py b/src/pip/_internal/legacy_resolve.py
index c24158f..bd92287 100644
--- a/src/pip/_internal/legacy_resolve.py
+++ b/src/pip/_internal/legacy_resolve.py
@@ -30,6 +30,7 @@ from pip._internal.exceptions import (
 )
 from pip._internal.utils.logging import indent_log
 from pip._internal.utils.misc import (
+    dist_in_install_path,
     dist_in_usersite,
     ensure_dir,
     normalize_version_info,
@@ -224,7 +225,9 @@ class Resolver(object):
         """
         # Don't uninstall the conflict if doing a user install and the
         # conflict is not a user install.
-        if not self.use_user_site or dist_in_usersite(req.satisfied_by):
+        if ((not self.use_user_site
+                or dist_in_usersite(req.satisfied_by))
+                and dist_in_install_path(req.satisfied_by)):
             req.conflicts_with = req.satisfied_by
         req.satisfied_by = None
 
diff --git a/src/pip/_internal/req/req_install.py b/src/pip/_internal/req/req_install.py
index 5a8c0dc..f80ba87 100644
--- a/src/pip/_internal/req/req_install.py
+++ b/src/pip/_internal/req/req_install.py
@@ -39,6 +39,7 @@ from pip._internal.utils.misc import (
     ask_path_exists,
     backup_dir,
     display_path,
+    dist_in_install_path,
     dist_in_site_packages,
     dist_in_usersite,
     ensure_dir,
@@ -461,7 +462,7 @@ class InstallRequirement(object):
                         "lack sys.path precedence to %s in %s" %
                         (existing_dist.project_name, existing_dist.location)
                     )
-            else:
+            elif dist_in_install_path(existing_dist):
                 self.conflicts_with = existing_dist
         return True
 
diff --git a/src/pip/_internal/utils/misc.py b/src/pip/_internal/utils/misc.py
index b848263..5b75fed 100644
--- a/src/pip/_internal/utils/misc.py
+++ b/src/pip/_internal/utils/misc.py
@@ -28,6 +28,7 @@ from pip._vendor.six.moves.urllib.parse import unquote as urllib_unquote
 from pip import __version__
 from pip._internal.exceptions import CommandError
 from pip._internal.locations import (
+    distutils_scheme,
     get_major_minor_version,
     site_packages,
     user_site,
@@ -389,6 +390,16 @@ def dist_in_site_packages(dist):
     return dist_location(dist).startswith(normalize_path(site_packages))
 
 
+def dist_in_install_path(dist):
+    """
+    Return True if given Distribution is installed in
+    path matching distutils_scheme layout.
+    """
+    norm_path = normalize_path(dist_location(dist))
+    return norm_path.startswith(normalize_path(
+        distutils_scheme("")['purelib'].split('python')[0]))
+
+
 def dist_is_editable(dist):
     # type: (Distribution) -> bool
     """
-- 
2.20.1

